src/components/include/sl.h

```diff
@@ -482,16 +494,8 @@ sl_cs_exit_schedule_nospin_arg(struct sl_thd *to)
 	 * the inter-component delegations), block till next timeout and try again.
 	 */
 	if (unlikely(ret == -EPERM)) {
-			cycles_t abs_timeout = globals->timer_next;
-
-			assert(t != globals->sched_thd);
-
-			sl_cs_enter();
-			if (likely(t->period)) abs_timeout = t->last_replenish + t->period;
-			sl_thd_block_no_cs(t, SL_THD_BLOCKED_TIMEOUT, abs_timeout);
-			sl_cs_exit();
-
-			if (unlikely(sl_thd_curr() != globals->sched_thd)) ret = sl_thd_activate(globals->sched_thd, tok);
+		sl_thd_block_expiry(t);
+		if (unlikely(sl_thd_curr() != globals->sched_thd)) ret = sl_thd_activate(globals->sched_thd, tok);
```

Comment:
Thanks for pointing this out. 
I think I thought about it but I wasn't sure what we'd do if this fails because we're trying to switch to the `sched_thd` here. 
We could be call a thread switch API `cos_thd_switch()` that doesn't take a TCAP argument and switch to the scheduler thread or `cos_switch()` by not passing a tcap argument. I think that might be better.